### PR TITLE
Wide-deep notebook bug fix

### PR DESCRIPTION
**Pull request by loomlike**
### Description
`ITEM_FEAT_COL=None` error fix
Ensure train at least 1 batch (when Epochs=0)
azureml_hyperdrive_wide_and_deep.ipynb link fix
Refactor dataloading and splitting cells
Refactor tests to use tempfile

### Related Issues
#685

### Checklist:

- [x] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING.md).
- [x] I have added tests.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: tests/integration/test_notebooks_gpu.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
-import shutil
+from tempfile import TemporaryDirectory
```
Comment (yueguoguo) [inline tests/integration/test_notebooks_gpu.py]: Interestingly, when I do `pytest` with `-s` to print something to console it failed if `TemporaryDiretory` or `tempdir` is used; `gettempdir()` works.... 😂

Comment (miguelgfierro) [reply tests/integration/test_notebooks_gpu.py]: does this work for you?
```
with TemporaryDirectory() as tmp_dir:

```

Comment (loomlike) [reply tests/integration/test_notebooks_gpu.py]: lol ... w/o '-s' does this work?

**Comment by miguelgfierro** [comment]
@loomlike do you know why the tests are not being executed?

File: tests/unit/test_notebooks_gpu.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
-import shutil
+from tempfile import TemporaryDirectory
```
Comment (miguelgfierro) [inline tests/unit/test_notebooks_gpu.py]: for the tests, we could use the pytest fixture tmpdir https://docs.pytest.org/en/latest/tmpdir.html. It's quite neat. There is also tmpdir_factory available in case we need to have more control on the temporary folders

Comment (loomlike) [reply tests/unit/test_notebooks_gpu.py]: Good! I'm working on it.

**Comment by loomlike** [comment]
@miguelgfierro no idea why the first commit didn't trigger the tests lol. But the second commit went through all of them.

File: tests/integration/test_notebooks_gpu.py (hunk 1 above)
Comment (yueguoguo) [reply tests/integration/test_notebooks_gpu.py]: Neither of the above worked for me... `pytest` fixture `tmpdir` works, so this is so far the best practice.

**Comment by anargyri** [comment]
> This looks nice, just one question that I asked you in the other PR, is `str(tmpdir.mkdir("wide_deep_0"))` the recommended way of getting the string of the path?

Also worth noting that they use a python library "not recommended for new projects":
https://docs.pytest.org/en/latest/tmpdir.html
https://py.readthedocs.io/en/latest/path.html

File: notebooks/00_quick_start/wide_deep_movielens.ipynb [hunk 3]
```diff
@@ -47,24 +47,23 @@
     "import sys\n",
```
Comment (anargyri) [inline notebooks/00_quick_start/wide_deep_movielens.ipynb]: <p>You don't need MODEL_DIR = None</p><p>since you do</p><p>if MODEL_DIR is None:</p><p>&nbsp;&nbsp;tmp_dir = TemporaryDirectory()</p><p>&nbsp;&nbsp;MODEL_DIR = tmp_dir.name</p><p>afterwards.</p>

 _Reply via <a href='https://app.reviewnb.com/Microsoft/Recommenders/pull/686/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (anargyri) [inline notebooks/00_quick_start/wide_deep_movielens.ipynb]: <p>You say "the hyperparameters are&nbsp;optimized for the training set."</p><p>But for the tuning there is also a validation set.</p>

 _Reply via <a href='https://app.reviewnb.com/Microsoft/Recommenders/pull/686/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Comment by anargyri** [comment]
Part of this notebook is about preprocessing and encoding features. I suspect this can be used by other algorithms in the future, so it is probably better to separate it into a util instead.

File: notebooks/00_quick_start/wide_deep_movielens.ipynb (hunk 3 above)
Comment (loomlike) [reply notebooks/00_quick_start/wide_deep_movielens.ipynb]: good catch. more specifically, we further split the training set into training / validation (instead of split the data into three from the first place). The reason was to make sure azureml_w_d notebook uses the same set of training set as the other notebooks. I will put more contextual detail to this notebook.

Comment (loomlike) [reply notebooks/00_quick_start/wide_deep_movielens.ipynb]: I will get `name 'MODEL_DIR' is not defined` error if I don't define it. Right?

**Comment by loomlike** [comment]
@anargyri thanks for great suggestions! See my response here:
* preprocessing and encoding features: totally agree. Actually, I'm planning to work on utilizing other features as well, like **movie's released year vs rating timestamp** (assuming users rated movies by the time they "watched"). Let me create another issue for that and take care "adding util functions" from there.
* pytest's tmpdir fixture: Yeah pytest is using py.path. I think in the future version of pytest, they will (should) update the underneath module to pathlib, and we can keep using as it is (hopefully). Any suggestions?

**Comment by loomlike** [comment]
@miguelgfierro Regarding `tmpdir`, here, I will get this error if I pass the object directly:
`Object of type 'LocalPath' is not JSON serializable`

I will update this accordingly once we have the conclusion from the other thread!

File: notebooks/00_quick_start/wide_deep_movielens.ipynb (hunk 3 above)
Comment (anargyri) [reply notebooks/00_quick_start/wide_deep_movielens.ipynb]: I mean, you just need 
`tmp_dir = TemporaryDirectory()
MODEL_DIR = tmp_dir.name`
You don't need the if statement because there is no alternative.

**Comment by anargyri** [comment]
> * pytest's tmpdir fixture: Yeah pytest is using py.path. I think in the future version of pytest, they will (should) update the underneath module to pathlib, and we can keep using as it is (hopefully). Any suggestions?

See my comment in another PR; there is a factory that uses pathlib
https://docs.pytest.org/en/latest/reference.html#tmp-path-factory-factory-api

File: tests/integration/test_notebooks_gpu.py (hunk 1 above)
Comment (loomlike) [reply tests/integration/test_notebooks_gpu.py]: @yueguoguo Can you double check what was the reason for the failure? @miguelgfierro and I had discussed and we want to use `with TemporaryDirectory() as td:` back again, because:

pytest's fixture (tmpdir or tmp_path or tmp_factory) keeps 3-latest temp folders and we don't want that (also saw someone complaining about the old folders are not removed completely from their CI machine).

I tested using TemporaryDirectory from my side and both unit and smoke tests are passed, with `-s`

Comment (yueguoguo) [reply tests/integration/test_notebooks_gpu.py]: Tested and it works with the context manager. Thanks!

File: notebooks/00_quick_start/wide_deep_movielens.ipynb (hunk 3 above)
Comment (loomlike) [reply notebooks/00_quick_start/wide_deep_movielens.ipynb]: Ah, I got your point. MODEL_DIR can be passed as a parameter when the notebook got called by papermill. Usecase is to store model checkpoints.
Will add comments to make it clear.

File: tests/integration/test_notebooks_gpu.py (hunk 1 above)
Comment (loomlike) [reply tests/integration/test_notebooks_gpu.py]: Let me refactor all tempdir things in a new issue and PR! I will merge the current PR as soon as possible since the main fix is the bugfix.
